### PR TITLE
3 add sleep mode implementation

### PR DESCRIPTION
**Pull request by kylzrnr**

### CODE REVIEW

File: src/die_init.cpp [hunk 1]
```diff
@@ -341,10 +341,14 @@ namespace Die
...
                 // Get contents of reset register to check why we're booting
+                // If we fail to read reason register, continue as if normal boot
                 uint32_t reason = 0;
                 ret_code_t ret = sd_power_reset_reason_get(&reason);
-                ret = sd_power_reset_reason_clr(0xF000F);
-                bool fromSleep = (reason >> 16) == 1;
+                bool fromSleep = false;
+                if (ret == NRF_SUCCESS) {
+                    sd_power_reset_reason_clr(0xF000F);
+                    fromSleep = (reason >> 16) == 1;
```
Comment (jeansimonet) [inline src/die_init.cpp:350]: This is generally not safe. I would replace with a specific test against the flag you are looking for (i.e. (reason & 0xXXXXXXXX) != 0).

Comment (kylzrnr) [reply src/die_init.cpp]: Pushed a change that corrects this issue.

**Review by jeansimonet — CHANGES_REQUESTED** [review]

File: src/die_init.cpp [hunk 2]
```diff
@@ -340,8 +340,15 @@ namespace Die
                 initMainLogic();
                 NRF_LOG_INFO("---------------");
 
+                // Get contents of reset register to check why we're booting
+                uint32_t reason = 0;
+                ret_code_t ret = sd_power_reset_reason_get(&reason);
+                ret = sd_power_reset_reason_clr(0xF000F);
+                bool fromSleep = (reason >> 16) == 1;
+
```
[the remark this answers is not in the document]
Comment (kylzrnr) [reply src/die_init.cpp]: Pushed a new commit to fix this.

**Comment by kylzrnr** [comment]
All current requests have been addressed in the latest commit!

File: src/drivers_nrf/power_manager.cpp [hunk 3]
```diff
@@ -87,6 +87,19 @@ namespace PowerManager
         nrf_pwr_mgmt_shutdown(NRF_PWR_MGMT_SHUTDOWN_RESET);
     }
 
+    bool checkFromSleep() {
+        // Get contents of reset register to check why we're booting
+        // If we fail to read reason register, continue as if normal boot
```
Comment (jeansimonet) [inline src/drivers_nrf/power_manager.cpp:92]: Good job with the comment!

File: src/modules/behavior_controller.cpp [hunk 4]
```diff
@@ -81,7 +81,10 @@ namespace BehaviorController
                     // if (BatteryController::getCurrentLevel() > BATT_TOO_LOW_LEVEL) {
                         NRF_LOG_INFO("Triggering a HelloGoodbye Condition");
```
Comment (jeansimonet) [inline src/modules/behavior_controller.cpp:83]: Actually I like that you had the alternate log_info indicating we skipped the hellogoobye!

**Review by jeansimonet — APPROVED** [review]

File: src/modules/behavior_controller.cpp (hunk 4 above)
Comment (kylzrnr) [reply src/modules/behavior_controller.cpp:83]: I see. I can move that inside the conditional then.

Comment (kylzrnr) [reply src/modules/behavior_controller.cpp]: Done!

**Review by jeansimonet — APPROVED** [review]

**Pull request merged**